### PR TITLE
Problem: zmq_connect fails after disconnect due to RECONNECT_IVL == -1

### PR DESCRIPTION
**Pull request by bluca**
Solution: when a connection breaks and ZMQ_RECONNECT_IVL is set to -1,
which means a reconnection will not be attempted, send a message from
the I/O thread to the application thread to make the socket call
term_endpoint, which is the equivalent of manually calling
zmq_disconnect.
This way subsequent zmq_connect call to the same endpoint will attempt
again to do a connection.
Otherwise, for some socket types like SUBs, those new connects will
fail as the endpoint is recorded, despite the connection having been
permanently closed.

Add test cases to exercise this corner case with TCP and IPC.

### CODE REVIEW

**Comment by sigiesec** [comment]
@bluca Do you think this fix is related to #1590 and #883?

**Comment by bluca** [comment]
Mmh I don't think so, this is specifically related to automated disconnects (or failed connections) and that socket option

**Comment by sigiesec** [comment]
ok, just saw the issues again, and they seemed superficially related

File: src/session_base.cpp [hunk 1]
```diff
@@ -536,6 +536,11 @@ void zmq::session_base_t::reconnect ()
     //  Reconnect.
     if (options.reconnect_ivl != -1)
         start_connecting (true);
+    else {
+        std::string *ep = new (std::string);
```
Comment (a4z) [inline src/session_base.cpp]: when/how will this string be deleted?

File: src/command.hpp [hunk 2]
```diff
@@ -153,6 +155,12 @@ namespace zmq
             struct {
             } term_ack;
 
+            //  Sent by session_base (I/O thread) to socket (application thread)
+            //  to ask to disconnect the endpoint.
+            struct {
+                std::string *endpoint;
```
Comment (a4z) [inline src/command.hpp]: I think this should simply be a std::string, and maybe not a pointer to a string.
It seem to me it  is very unclear who is responsible for this pointer

File: src/session_base.cpp (hunk 1 above)
Comment (bluca) [reply src/session_base.cpp]: By the application thread on receipt of the message

File: src/command.hpp (hunk 2 above)
Comment (bluca) [reply src/command.hpp]: It must be either a pointer or a basic type, as it is part of a union

File: src/session_base.cpp (hunk 1 above)
Comment (bluca) [reply src/session_base.cpp]: The pointer is stored in the union args_t which is part of struct command_t, and the receiver deletes it here:

https://github.com/zeromq/libzmq/pull/2756/files#diff-4243b1744be5585239d76c7e7d18017fR1428
